platform/o.n.bootstrap/src/org/netbeans/JarClassLoader.java

```diff
@@ -574,13 +609,81 @@ protected URL doGetResource(String name) throws IOException  {
         @Override
         protected byte[] readClass(String path) throws IOException {
             try {
+                if (isMultiRelease() && RUNTIME_VERSION > BASE_VERSION) {
+                    int[] vers = getVersions();
+                    if (vers.length > 0) {
+                        for (int i = vers.length - 1; i >= 0; i--) {
+                            int version = vers[i];
+                            if (version > RUNTIME_VERSION) {
+                                continue;
+                            }
+                            if (version < BASE_VERSION) {
+                                break;
+                            }
+                            byte[] data = archive.getData(this, "META-INF/versions/" + version + "/" + path);
+                            if (data != null) {
+                                return data;
+                            }
+                        }
+                    }
+                }
                 return archive.getData(this, path);
             } catch (ZipException ex) {
                 dumpFiles(file, -1);
                 throw ex;
             }
         }
-        
+
+        private int[] getVersions() {
+            int[] ret;
+            if (versions != null && (ret = versions.get()) != null) {
+                return ret;
+            }
+            try {
+                JarFile src = getJarFile("versions");
+                Set<Integer> vers = new TreeSet<>();
+                Enumeration<JarEntry> en = src.entries();
+                while (en.hasMoreElements()) {
+                    JarEntry je = en.nextElement();
+                    if (je.isDirectory()) {
+                        String itm = je.getName();
+                        if (itm.startsWith("META-INF/versions/")) {
+                            String res = itm.substring(18);
+                            int idx = res.indexOf('/');
+                            if (idx > 0 && idx == res.length() - 1) {
+                                vers.add(Integer.parseInt(res.substring(0, idx)));
```

Comment:
would recommend to catch and ignore `NumberFormatException` in case something unknown shows up (defensive forward compatibility).

i converted this section to streams to check if it improves readability, I let you decide what you prefere:
```java
    int[] ret = getJarFile(path).stream()
        .filter(JarEntry::isDirectory)
        .filter((e) -> e.getName().startsWith("META-INF/versions/"))
        .mapToInt((e) -> {
            String res = e.getName().substring(18);
            int idx = res.indexOf('/');
            if (idx > 0 && idx == res.length() - 1) {
                try {
                    return Integer.parseInt(res.substring(0, idx));
                } catch (NumberFormatException ignore) {}
            }
            return -1;
        })
        .sorted().distinct().filter((v) -> v != -1).toArray();
```